/**
 *  Update API
 */

syntax = "proto3";

package peloton.api.v0.update;

option go_package = "peloton/api/v0/update";
option java_package = "peloton.api.v0.update";

import "peloton/api/v0/peloton.proto";

/**
 *  Update options for a job update
 */
message UpdateConfig {
  // Update batch size of the deployment
  uint32 batchSize = 1;

  // Update batch percentage of the deployment. If present,
  // will take precedence over batchSize
  double batchPercentage = 2;

  // Whether or not to stop all instance before update
  bool stopBeforeUpdate = 3;

  // startPaused indicates if the update should start in the paused state,
  // requiring an explicit resume to initiate.
  bool startPaused = 4;

  // rollbackOnFailure indicates if the update should be rolled back
  // automatically if failure is detected
  bool rollbackOnFailure = 5;

  // maxInstanceAttempts is the maximum attempt times for one task.
  // If the value is 0, the instance can be retried for infinite
  // times.
  uint32 maxInstanceAttempts = 6;
  
  // maxFailureInstances is the number of failed instances in one
  // update that allowed.
  // If the value is 0, there is no limit for max failure instances
  // and the update is marked successful even if all of the instances
  // fail.
  uint32 maxFailureInstances = 7;

  // If set to true, peloton would try to place the task
  // restarted/updated on the host it previously run on.
  // It is best effort, and has no guarantee of success.
  bool inPlace = 8;

  // Experimental: this flag is for experimental purpose and
  // will be removed in the future release.
  // If set to true, peloton would start tasks updated
  // if the tasks are killed previously.
  // By default, killed tasks would remain killed, and
  // run with new version when running again.
  bool startTasks = 9;
}

// Runtime state of a job update
enum State {
  // Invalid protobuf value
  INVALID = 0;

  // The update has been created but not started yet
  INITIALIZED = 1;

  // The update is rolling forward
  ROLLING_FORWARD = 2;

  // The update has been paused
  PAUSED = 3;

  // The update has completed successfully
  SUCCEEDED = 4;

  // The update was aborted/cancelled
  ABORTED = 5;

  // The update is failed
  FAILED = 6;

  // The update is rolling barckward
  ROLLING_BACKWARD = 7;

  // The update was rolled back due to failure
  ROLLED_BACK = 8;
}

/**
 *  UpdateStatus provides current runtime status of an update
 */
message UpdateStatus {
  // Number of tasks that have been updated
  uint32 numTasksDone = 1;

  // Number of tasks to be updated
  uint32 numTasksRemaining = 2;

  // Runtime state of the update
  State state = 3;

  // Number of tasks that failed during the update
  uint32 numTasksFailed = 4;
}

/**
 * Information of an update, such as update config and runtime status
 */
message UpdateInfo {
  // Update ID of the job update
  peloton.UpdateID updateId = 1;

  // Update configuration
  UpdateConfig config = 2;

  // Update runtime status
  UpdateStatus status = 3;

  // Job ID of the job update
  peloton.JobID jobId = 4;

  // Configuration version of the job after this update
  uint64 configVersion = 5;

  // Job configuration version before the update
  uint64 prevConfigVersion = 6;

  // Opaque metadata provided by the user
  peloton.OpaqueData opaque_data = 7;
}
